include/deal.II/grid/cell_id.h

```diff
@@ -48,34 +52,61 @@ template <int, int> class Triangulation;
  *
  * @note How this data is internally represented is not of importance (and not
  * exposed on purpose).
- *
- * @todo Does it make sense to implement a more efficient representation
- * (internally and/or as a string)? If yes, something like a 64bit int as in
- * p4est would be a good option.
  */
 class CellId
 {
 public:
   /**
-   * Construct a CellId object with a given @p coarse_cell_index and list of child indices.
+   * A type that is used to encode the CellId data in a compact and fast way
+   * (e.g. for MPI transfer to other processes). Note that it limits the
+   * number of children that can be transferred to 20 in 3D and 30 in 2D
+   * (using 2 times 32 bit for storage), a limitation that is identical to
+   * the one used by p4est.
+   */
+  typedef std_cxx11::array<unsigned int, 4> binary_type;
+
+  /**
+   * Construct a CellId object with a given @p coarse_cell_index and vector of
+   * child indices.
    */
   CellId(const unsigned int coarse_cell_id,
          const std::vector<unsigned char> &child_indices);
 
   /**
+   * Construct a CellId object with a given @p coarse_cell_index and array of
+   * child indices provided in @p child_indices. @p child_indices shall have
+   * at least @p n_child_indices valid entries.
+   */
+  CellId(const unsigned int coarse_cell_id,
+         const unsigned int n_child_indices,
+         const unsigned char *child_indices);
+
+  /**
+   * Construct a CellId object with a given binary representation that was
+   * previously constructed by CellId::to_binary.
+   */
+  CellId(const binary_type &binary_representation);
+
+  /**
    * Construct an invalid CellId.
    */
   CellId();
 
   /**
-   * Return a string representation of this CellId.
+   * Return a human readable string representation of this CellId.
    */
   std::string to_string() const;
 
   /**
+   * Return a compact and fast binary representation of this CellId.
+   */
+  template<int dim>
+  binary_type to_binary() const;
+
+  /**
    * Return a cell_iterator to the cell represented by this CellId.
    */
-  template<int dim, int spacedim>
+  template<int dim, int spacedim=dim>
   typename Triangulation<dim,spacedim>::cell_iterator
   to_cell(const Triangulation<dim,spacedim> &tria) const;
```

Comment:
the `=dim` in the template arg list requires C++11. But you don't need the default value -- the compiler can deduce both template types.